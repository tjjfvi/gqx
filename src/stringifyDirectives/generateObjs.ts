
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
