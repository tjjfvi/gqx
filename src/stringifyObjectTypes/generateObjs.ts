
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
