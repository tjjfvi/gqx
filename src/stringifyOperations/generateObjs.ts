
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
