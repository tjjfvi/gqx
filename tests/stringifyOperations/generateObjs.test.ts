import generateObjs from "../../src/stringifyOperations/generateObjs";

test("", () => {
  const [objs] = generateObjs(utils.getOperations(["extensive"]));
  expect(objs.map(o => ({ n: o.type, p: o.props.length }))).toMatchSnapshot();
  expect(objs).toMatchSnapshot();
});
