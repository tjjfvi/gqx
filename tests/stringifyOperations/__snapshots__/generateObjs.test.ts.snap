// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "n": "Query",
    "p": 3,
  },
  Object {
    "n": "Mutation",
    "p": 1,
  },
]
`;

exports[` 2`] = `
Array [
  Object {
    "prop": "query",
    "props": Array [
      Object {
        "args": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "kind": "InputValueDefinition",
            "loc": Object {
              "end": 82,
              "start": 75,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 77,
                "start": 75,
              },
              "value": "id",
            },
            "type": Object {
              "kind": "NonNullType",
              "loc": Object {
                "end": 82,
                "start": 79,
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 81,
                  "start": 79,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 81,
                    "start": 79,
                  },
                  "value": "ID",
                },
              },
            },
          },
        ],
        "id": Object {
          "prop": "getAuthor",
          "type": "Query",
        },
        "type": "Author",
        "wrap": [Function],
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "kind": "InputValueDefinition",
            "loc": Object {
              "end": 110,
              "start": 103,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 105,
                "start": 103,
              },
              "value": "id",
            },
            "type": Object {
              "kind": "NonNullType",
              "loc": Object {
                "end": 110,
                "start": 107,
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 109,
                  "start": 107,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 109,
                    "start": 107,
                  },
                  "value": "ID",
                },
              },
            },
          },
        ],
        "id": Object {
          "prop": "getBook",
          "type": "Query",
        },
        "type": "Book",
        "wrap": [Function],
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "kind": "InputValueDefinition",
            "loc": Object {
              "end": 145,
              "start": 131,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 137,
                "start": 131,
              },
              "value": "cursor",
            },
            "type": Object {
              "kind": "NamedType",
              "loc": Object {
                "end": 145,
                "start": 139,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 145,
                  "start": 139,
                },
                "value": "Cursor",
              },
            },
          },
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "kind": "InputValueDefinition",
            "loc": Object {
              "end": 165,
              "start": 147,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 153,
                "start": 147,
              },
              "value": "filter",
            },
            "type": Object {
              "kind": "NamedType",
              "loc": Object {
                "end": 165,
                "start": 155,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 165,
                  "start": 155,
                },
                "value": "BookFilter",
              },
            },
          },
        ],
        "id": Object {
          "prop": "listBooks",
          "type": "Query",
        },
        "type": "Book",
        "wrap": [Function],
      },
    ],
    "type": "Query",
  },
  Object {
    "prop": "mutation",
    "props": Array [
      Object {
        "args": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "kind": "InputValueDefinition",
            "loc": Object {
              "end": 227,
              "start": 208,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 213,
                "start": 208,
              },
              "value": "input",
            },
            "type": Object {
              "kind": "NonNullType",
              "loc": Object {
                "end": 227,
                "start": 215,
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 226,
                  "start": 215,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 226,
                    "start": 215,
                  },
                  "value": "ReviewInput",
                },
              },
            },
          },
        ],
        "id": Object {
          "prop": "addReview",
          "type": "Mutation",
        },
        "type": "Review",
        "wrap": [Function],
      },
    ],
    "type": "Mutation",
  },
]
`;
